archive/multisource.go

```diff
@@ -0,0 +1,174 @@
+package archive
+
+import (
+	"context"
+	"io"
+
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/filter"
+	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/scanner"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+// A spanInfo is a logical view of the records within a time span, stored
+// in one or more Chunks.
+type spanInfo struct {
+	span nano.Span
+
+	// chunks are the data files that contain records within this spanInfo.
+	// The Chunks may have spans that extend beyond this spanInfo, so any
+	// records from these Chunks should be limited to those that fall within a
+	// closed span constructed from First & Last.
+	chunks []Chunk
+}
+
+func spanWalk(ctx context.Context, ark *Archive, filter nano.Span, v func(si spanInfo) error) error {
+	return tsDirVisit(ctx, ark, filter, func(_ tsDir, chunks []Chunk) error {
+		sinfos := mergeChunksToSpans(chunks, ark.DataSortDirection, filter)
+		for _, s := range sinfos {
+			if err := v(s); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+type multiCloser struct {
+	closers []io.Closer
+}
+
+func (c *multiCloser) Close() error {
+	var err error
+	for _, c := range c.closers {
+		if closeErr := c.Close(); err == nil {
+			err = closeErr
+		}
+	}
+	return err
+}
+
+type scannerCloser struct {
+	scanner.Scanner
+	io.Closer
+}
+
+func newSpanScanner(ctx context.Context, ark *Archive, zctx *resolver.Context, f filter.Filter, filterExpr ast.BooleanExpr, si spanInfo) (sc *scannerCloser, err error) {
+	if len(si.chunks) == 1 {
+		rc, err := iosrc.NewReader(ctx, si.chunks[0].Path(ark))
+		if err != nil {
+			return nil, err
+		}
+		sn, err := scanner.NewScanner(ctx, zngio.NewReader(rc, zctx), f, filterExpr, si.span)
+		if err != nil {
+			rc.Close()
+			return nil, err
+		}
+		return &scannerCloser{sn, rc}, nil
+	}
+	closers := make([]io.Closer, len(si.chunks))
+	defer func() {
+		if err != nil {
+			for _, c := range closers {
+				c.Close()
+			}
+		}
+	}()
+	readers := make([]zbuf.Reader, len(si.chunks))
+	for i, chunk := range si.chunks {
+		rc, err := iosrc.NewReader(ctx, chunk.Path(ark))
+		if err != nil {
+			return nil, err
+		}
+		closers[i] = rc
+		readers[i] = zngio.NewReader(rc, zctx)
+	}
+	sn, err := scanner.NewCombiner(ctx, readers, zbuf.RecordCompare(ark.DataSortDirection), f, filterExpr, si.span)
```

Comment:
This is a bit of an edge case, but if want raw, in-order searches to be deterministic shouldn't the `filterExpr` passed into the Combiner here take into account the chunk id, so it can consistently sort equal timestamps from separate chunks?

Reply:
You're right; this another instance of this issue: https://github.com/brimsec/zq/issues/1215#issuecomment-698356017 . As mentioned there, a useful technique is to use the zcode.Bytes of the record itself to break sort field ties.

Reply:
Even better, this wouldn't include the complication of having to some how include chunkid in the combiner.